feat: Authenticate private resource requests

dcalhoun opened the pull request:
## Related

- CMM-235-linear-issue
- https://github.com/Automattic/dotcom-forge/issues/9446
- https://github.com/wordpress-mobile/GutenbergKit/pull/34
- https://github.com/wordpress-mobile/GutenbergKit/pull/30

## Description

Authenticate resource requests originating from within the WebView so that
requests to private sites succeed. Ref CMM-235.

## To Test:

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

> [!Important]
> The proposed logic does not work for streamed files (e.g., video and audio). It appears [`shouldInterceptRequest` injects a `Content-Length: 0` header](https://groups.google.com/a/chromium.org/g/net-dev/c/vcglzJAXwaU/m/itqvpU7qHgAJ), which creates duplicate values creating a malformed request that disables streaming. 

See https://github.com/wordpress-mobile/GutenbergKit/pull/34.

## Regression Notes

1. Potential unintended areas of impact
    Gutenberg Mobile or Aztec editors fail. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manually tested both editors. 
3. What automated tests I added (or what prevented me from doing so)
    None, feels unnecessary for this currently experimental editor feature. 

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1635,4 +1656,71 @@ public void onConnectionStatusChange(boolean isConnected) {
+
+        String proxyUrl = url.toString();
+        if (mIsPrivateAtomic) {
+            proxyUrl = getPrivateResourceProxyUrl(url);
```
Private Atomic sites require a proxy for request authentication (D38925-code).

jkmassel commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1635,4 +1656,71 @@ public void onConnectionStatusChange(boolean isConnected) {
+        Uri url = request.getUrl();
+        String siteURL = (String) (mSettings != null ? mSettings.get("siteURL") : "");
+        String siteHostedMedia = siteURL + "/.*\\.(jpg|jpeg|png|gif|bmp|webp|mp4|mov|avi|mkv|mp3|wav|flac)(\\?.*)?$";
+        Pattern pattern = Pattern.compile(siteHostedMedia, Pattern.CASE_INSENSITIVE);
```
I'm not sure a regex is the way to do this? URL parsing is probably more performant and accurate. WDYT?

dcalhoun replied:
Makes sense to me. I was merely unfamiliar with the built-in URL parsing. I'll refactor this.

dcalhoun replied:
Refactored in a9fe1d56d3b4b7c5796337f27c463095400a5ddb.

dcalhoun commented:
👋🏻 @jkmassel. I updated this with the merged https://github.com/wordpress-mobile/GutenbergKit/pull/30 and the latest from `trunk`. I re-tested and it appears to be functioning as expected. Whenever you have time, we can look to review and merge this feature. Thanks!

nbradbury commented:
@dcalhoun @jkmassel Where do we stand with this PR? Is it something we can close?

dcalhoun commented:
> Where do we stand with this PR? Is it something we can close?

Ideally, it is merged. 

However, @jkmassel mentioned privately he would prefer avoiding a [proxy URL](https://github.com/wordpress-mobile/WordPress-Android/pull/21331#discussion_r1822913488). Transparently, I was unsure as to how we might accomplish that, so I have not moved this forward. @jkmassel do you have specific ideas for avoiding the proxy URL?

dcalhoun committed:
feat: Authenticate private resource requests

Unauthenticated requests fail when attempting to load resources from a
private site--e.g., an image.

dcalhoun committed:
style: Address lint warnings

Avoid NullPointerExceptions.

dcalhoun committed:
style: Address "function too long" lint warning

Lint configuration warned the `createGutenbergKitEditorFragment`
function was over 60 lines.

dcalhoun committed:
feat: Expand supported media types

We should authenticate requests for these common media types.
